### DISCUSSION TITLE
Why are calls to virtual generic methods so much slower than calls to other methods?

### DISCUSSION THREAD

**Discussion by dczuo** [dotnet/runtime#65767] [Q&A]
```c#
[Benchmark]
public void InvokeM()
{
	M();
}

[Benchmark]
public void InvokeVM()
{
	VM();
}

[Benchmark]
public void InvokeGM()
{
	GM<string>();
}

[Benchmark]
public void InvokeVGM()
{
	VGM<string>();
}

void M()
{
}

protected virtual void VM()
{
}

void GM<T>()
{
}

protected virtual void VGM<T>()
{
}
```
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1566 (21H2)
Intel Core i7-3520M CPU 2.90GHz (Ivy Bridge), 1 CPU, 4 logical and 2 physical cores
.NET SDK=6.0.200
  [Host]     : .NET 6.0.2 (6.0.222.6406), X64 RyuJIT  [AttachedDebugger]
  DefaultJob : .NET 6.0.2 (6.0.222.6406), X64 RyuJIT

|    Method |       Mean |     Error |    StdDev |     Median | Allocated |
|---------- |-----------:|----------:|----------:|-----------:|----------:|
|   InvokeM |  0.0096 ns | 0.0152 ns | 0.0142 ns |  0.0000 ns |         - |
|  InvokeVM |  0.6099 ns | 0.0160 ns | 0.0150 ns |  0.6129 ns |         - |
|  InvokeGM |  0.0031 ns | 0.0074 ns | 0.0066 ns |  0.0000 ns |         - |
| InvokeVGM | 12.5043 ns | 0.1471 ns | 0.1376 ns | 12.5060 ns |         - |

**Comment by EgorBo** [comment]
First of all, your `InvokeM` and `InvokeGM` do not measure any call overhead - they're empty after inlining.

**Reply by dczuo** [reply]
> First of all, your `InvokeM` and `InvokeGM` do not measure any call overhead - they're empty after inlining.

Thanks for the quick reply! What about InvokeVM and InvokeVGM?

**Reply by EgorBo** [reply]
`InvokeVM` is a pretty fast virtual call, can be optimized if you turn on Dynamic PGO, see https://gist.github.com/EgorBo/dc181796683da3d905a5295bfd3dd95b

Virtual Generics are known to be slow 😞

**Reply by EgorBo** [reply]
![image](https://user-images.githubusercontent.com/523221/155345080-464243b7-5616-4c7b-9635-d3ed9c89c6c3.png)

cc @jkotas I wonder if we can improve dictionary lookup performance, e.g. .NET Dictionary uses "Fast Mod" algorithm while native one seems to be does not.
![image](https://user-images.githubusercontent.com/523221/155345011-41ffbd44-9468-44a1-806d-48316673d6c3.png)
(to avoid idiv with ~40 latency)

Also, it seems we don't hoist/CSE its arguments on jit side:

![image](https://user-images.githubusercontent.com/523221/155346177-ac7df49c-1285-487d-add9-39b10f00c19b.png)

**Reply by EgorBo** [reply]
Most of the perf seems to be spent around that `div`:
![image](https://user-images.githubusercontent.com/523221/155347571-03087d1d-984d-4a12-9957-a13a29802ec8.png)

**Reply by EgorBo** [reply]
Moved discussion to https://github.com/dotnet/runtime/issues/65778

**Reply by Xyncgas** [reply]
so this is how they rigged our election

**Reply by richlander** [reply]
Please keep the discourse on-topic and productive @Xyncgas.
